rust/src/passwd.rs

```diff
@@ -567,3 +634,299 @@ impl PasswdDB {
+    }
+
+    /// Add all users from a given `passwd` file.
+    pub fn add_passwd_content(&mut self, rootfs_dfd: i32, passwd_path: &str) -> Result<()> {
+        let rootfs = ffiutil::ffi_view_openat_dir(rootfs_dfd);
+        let db = rootfs.open_file(passwd_path)?;
+        let entries = nameservice::passwd::parse_passwd_content(BufReader::new(db))?;
+
+        for user in entries {
+            let uid = Uid::from_raw(user.uid);
+            let gid = Gid::from_raw(user.gid);
+            self.users.insert(user.name, (uid, gid));
+        }
+        Ok(())
+    }
+
+    /// Check whether the given username exists among user entries.
+    pub fn contains_user(&self, username: &str) -> bool {
+        self.users.contains_key(username)
+    }
+
+    /// Check whether the given groupname exists among group entries.
+    pub fn contains_group(&self, groupname: &str) -> bool {
+        self.groups.contains_key(groupname)
+    }
+
+    /// Lookup user ID by name.
+    pub fn lookup_user_id(&self, username: &str) -> Result<u32> {
+        self.users
+            .get(username)
+            .map(|user| user.0.as_raw())
+            .ok_or_else(|| anyhow!("failed to find user '{}'", username))
+    }
+
+    /// Lookup group ID by name.
+    pub fn lookup_group_id(&self, groupname: &str) -> Result<u32> {
+        self.groups
+            .get(groupname)
+            .map(|gid| gid.as_raw())
+            .ok_or_else(|| anyhow!("failed to find group '{}'", groupname))
+    }
+
+    #[context("Rendering user entries from treefile check-passwd")]
+    fn populate_users_from_treefile(
+        &mut self,
+        treefile: &mut Treefile,
+        repo_previous_rev: &Option<(&ostree::Repo, &str)>,
+    ) -> Result<()> {
+        let config = treefile.parsed.get_check_passwd();
+
+        match config {
+            CheckPasswd::None => {}
+            CheckPasswd::File(f) => {
+                let fp = treefile.externals.passwd_file_mut(f)?;
+                let buf_rd = BufReader::new(fp);
+                let entries = nameservice::passwd::parse_passwd_content(buf_rd)?;
+                for user in entries {
+                    self.users.insert(
+                        user.name,
+                        (Uid::from_raw(user.uid), Gid::from_raw(user.gid)),
+                    );
+                }
+            }
+            CheckPasswd::Previous => {
+                // This logic short-circuits if there is no previous commit, or if
+                // it doesn't contain a passwd file. Nothing to validate in that case.
+                let (repo, previous_rev) = match repo_previous_rev {
+                    Some(v) => v,
+                    None => return Ok(()),
+                };
+                let (prev_root, _name) = repo.read_commit(previous_rev, gio::NONE_CANCELLABLE)?;
+                let old_path = match prev_root.resolve_relative_path("usr/lib/passwd") {
+                    Some(v) => v,
+                    None => return Ok(()),
```

Comment:
Note: this case was not handled in the original C logic. I arbitrarily chose to short-circuit, but we could also decide to error out.